### PR TITLE
Parse symbolic shapes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69871
* #71170
* #71121
* #69776
* __->__ #69775
* #69645

Adds parsing for Symbolic Shapes.

Differential Revision: [D33515233](https://our.internmc.facebook.com/intern/diff/D33515233)

### CODE REVIEW

File: torch/csrc/jit/frontend/schema_type_parser.cpp [hunk 1]
```diff
@@ -206,7 +206,7 @@ TypePtr SchemaTypeParser::parseRefinedTensor() {
   std::vector<int64_t> strides;
   parseList(TK_NOTHING, ',', ')', [&] {
     // Extra handling for options like 'device' and 'requires_grad'
-    if (L.cur().kind == TK_IDENT) {
+    if (L.cur().kind == TK_IDENT && L.cur().text() != "SS") {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/frontend/schema_type_parser.cpp:209]: Nit: There is a comment in the beginning showing what parser supports, might be worth mentioning symbolic shapes in examples as well.

File: torch/csrc/jit/frontend/schema_type_parser.cpp [hunk 2]
```diff
@@ -257,10 +257,21 @@ TypePtr SchemaTypeParser::parseRefinedTensor() {
...
+      L.next();
+      L.expect('(');
+      L.expect('-');
+      shape_symbol = true;
+    }
     const std::string& num = L.expect(TK_NUMBER).text();
     // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
     std::string::size_type num_len;
-    auto dim = c10::stoll(num, &num_len);
+    int64_t dim = c10::stoll(num, &num_len);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/frontend/schema_type_parser.cpp:270]: Just to double check: when we pass `[10, -2, 30]` as `dims`, `TensorType::create` will figure out that it needs to create a symbolic shape for -2, right?

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

File: torch/csrc/jit/frontend/schema_type_parser.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/frontend/schema_type_parser.cpp:270]: yup

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33515233).

**Pull request closed without merging**